attest/win_events.go

```diff
@@ -0,0 +1,462 @@
+}
+
+func (w *WinEvents) readLoadedModuleAggregation(rdr io.Reader, header microsoftEventHeader) error {
+	var (
+		r        = io.LimitReader(rdr, int64(header.Size))
+		codeHash []byte
+		imgBase  uint64
+	)
+
+eventLoop:
+	for {
+		var header microsoftEventHeader
+		if err := binary.Read(r, binary.LittleEndian, &header); err != nil {
+			if errors.Is(err, io.EOF) {
+				break eventLoop
+			}
+			return fmt.Errorf("parsing LMA sub-event: %v", err)
+		}
+
+		var err error
+		switch header.Type {
+		case imageBase:
+			if imgBase != 0 {
+				return errors.New("duplicate image base data in LMA event")
+			}
+			if imgBase, err = w.parseImageBase(header, r); err != nil {
+				return err
+			}
+		case authenticodeHash:
+			if codeHash != nil {
+				return errors.New("duplicate authenticode hash structure in LMA event")
+			}
+			if codeHash, err = w.parseAuthenticodeHash(header, r); err != nil {
+				return err
+			}
+		default:
+			return fmt.Errorf("unknown event in LMA aggregation: %v", header.Type)
+		}
+	}
+
+	l := WinModuleLoad{
+		AuthenticodeHash: codeHash,
+		ImageBase:        []uint64{imgBase},
+	}
+	hashHex := hex.EncodeToString(l.AuthenticodeHash)
+	l.ImageBase = append(l.ImageBase, w.LoadedModules[hashHex].ImageBase...)
+	w.LoadedModules[hashHex] = l
+	return nil
+}
+
+func (w *WinEvents) readSIPAEvent(r io.Reader) error {
+	var header microsoftEventHeader
+	if err := binary.Read(r, binary.LittleEndian, &header); err != nil {
+		return err
+	}
+
+	// fmt.Printf("ID: %x (%v)\n", header.Type, header.Type&trustBoundary == trustBoundary)
+	switch header.Type {
+	case loadedModuleAggregation:
+		return w.readLoadedModuleAggregation(r, header)
+	case bootCounter:
+		return w.readBootCounter(header, r)
+	case bitlockerUnlock:
+		return w.readBitlockerUnlock(header, r)
+
+	case osKernelDebug, codeIntegrity, bootDebugging, testSigning: // Parse boolean values.
+		return w.readBooleanByteEvent(header, r)
+	case dataExecutionPrevention: // Parse booleans represented as uint64's.
+		return w.readBooleanInt64Event(header, r)
+
+	default:
+		// Event type was not handled, consume the data.
+		if header.Size > 16*1024 {
+			return fmt.Errorf("event data unreasonably large (%d bytes)", header.Size)
+		}
+		tmp := make([]byte, header.Size)
+		if err := binary.Read(r, binary.LittleEndian, &tmp); err != nil {
+			return fmt.Errorf("reading unknown data section of length %d: %w", header.Size, err)
+		}
+
+		return unknownSIPAEvent
+	}
+}
+
+// readWinEventBlock extracts boot configuration from SIPA events contained in
+// the given tagged event.
+func (w *WinEvents) readWinEventBlock(evt *internal.TaggedEventData) error {
+	r := bytes.NewReader(evt.Data)
+
+	// All windows information should be sub events in an enclosing SIPA
+	// container event.
+	if (windowsEvent(evt.ID) & sipaTypeMask) != sipaContainer {
+		return fmt.Errorf("expected container event, got %v", windowsEvent(evt.ID))
+	}
+
+eventLoop:
+	for {
+		if err := w.readSIPAEvent(r); err != nil {
```

Comment:
why isn't this

```
for r.Len() > 0 {
    if err := w.readSIPAEvent(r); err != nil {
        if errors.Is(err, unknownSIPAEvent) {
            // Unknown SIPA events are okay as all TCG events are verifiable.
            continue eventLoop
        }
        return err
    }
}
return nil
```

instead of doing error handling for EOF in the loop?